### PR TITLE
In a text widget, Font resources can be in the appearance

### LINKED ISSUE #12398: Font resources for a text widget can be in appearance
In file https://github.com/mozilla/pdf.js/files/5249872/product1.pdf, the fifth text widget has a default appearance using font F1 and this one is neither in DR nor in AcroForm but in AP::N::Resources.

### PR DESCRIPTION
**Pull request by calixteman**
In file https://github.com/mozilla/pdf.js/files/5249872/product1.pdf, the fifth text widget has a default appearance using font F1 and this one is neither in DR nor in AcroForm but in AP::N::Resources.

Fixes #12398.

### CODE REVIEW

**Comment by timvandermeij** [comment]
Is there test coverage for this or should a test be added? The file itself is already included in https://github.com/mozilla/pdf.js/pull/12393/files, but do we need a print test here?

**Comment by calixteman** [comment]
I can add a test.

**Comment by timvandermeij** [comment]
Looks good to me with a reference test added. Most likely this can be done in the same way as in https://github.com/mozilla/pdf.js/pull/12361/files.

File: src/core/annotation.js [hunk 1]
```diff
@@ -922,12 +922,17 @@ class WidgetAnnotation extends Annotation {
 
     const localResources = getInheritableProperty({ dict, key: "DR" });
     const acroFormResources = params.acroForm.get("DR");
+    const appearanceResources = this.appearance
+      ? this.appearance.dict.get("Resources")
+      : null;
+
```
Comment (Snuffleupagus) [inline src/core/annotation.js:928]: Do you actually need the ternary expression here?
```javascript
const appearanceResources =
  this.appearance && this.appearance.dict.get("Resources")`;
```
ought to suffice, given that all relevant code already contains `Dict` checks and having `appearanceResources` be `null` instead of `undefined` really shouldn't matter.

Comment (automatedbugreportingfacility) [reply src/core/annotation.js]: In the future, when all supported browsers handle optional chaining, this can be simplified further to:
```
const appearanceResources = this.appearance?.dict.get("Resources");
```

File: test/test_manifest.json [hunk 2]
```diff
@@ -4243,6 +4243,18 @@
        "lastPage": 1,
        "type": "eq"
     },
+    {  "id": "issue12394",
+       "file": "pdfs/issue12392.pdf",
```
Comment (timvandermeij) [inline test/test_manifest.json:4247]: This file is missing in the test folder. You'll need something similar to https://github.com/mozilla/pdf.js/pull/12361/files#diff-eba194d158906ef5ebd771295e400e6a.

**Comment by timvandermeij** [comment]
@calixteman If you have time, could you address the comments in this and the other PRs so we can get some of them merged?

File: test/test_manifest.json (hunk 2 above)
Comment (calixteman) [reply test/test_manifest.json:4247]: I'm reusing the test file for #12392:
https://github.com/mozilla/pdf.js/blob/master/test/pdfs/issue12392.pdf.link
So it should be ok, isn't it ? or need I to duplicate the pdf ?

Comment (Snuffleupagus) [reply test/test_manifest.json:4247]: > So it should be ok, isn't it ?

Of course it's fine to re-use one PDF file for multiple tests, we already do that in other tests.

File: test/test_manifest.json [hunk 3]
```diff
@@ -4243,6 +4243,18 @@
...
+       "md5": "76c3a34c6520940c45c66c92f7df2de5",
+       "rounds": 1,
+       "link": true,
+       "lastPage": 1,
+       "type": "eq",
+       "print": "true",
+       "annotationStorage": {
+           "35R": "收文"
```
Comment (Snuffleupagus) [inline test/test_manifest.json:4255]: Nit: A little bit too much indentation here.

File: src/core/annotation.js [hunk 4]
```diff
@@ -922,12 +922,16 @@ class WidgetAnnotation extends Annotation {
...
+      this.appearance && this.appearance.dict.get("Resources");
+
     this._fieldResources = {
       localResources,
       acroFormResources,
+      appearanceResources,
       mergedResources: Dict.merge({
         xref: params.xref,
-        dictArray: [localResources, acroFormResources],
+        dictArray: [localResources, acroFormResources, appearanceResources],
```
Comment (Snuffleupagus) [inline src/core/annotation.js:934]: Just a small questions here: Do you want this ordering, since that'd prefer the `appearanceResources` *last* when merging the dictionaries, since `Dict.merge` iterates through this array left-to-right?
Or would swapping the order of `acroFormResources` and `appearanceResources` here make sense?

It probably doesn't matter, since this situation probably is an an edge-case anyway, but figured I'd ask anyway :-)

Comment (calixteman) [reply src/core/annotation.js:934]: Probably a edge case, but I think it's better to swap.

File: test/test_manifest.json (hunk 2 above)
Comment (timvandermeij) [reply test/test_manifest.json:4247]: I completely missed this since no file was added and the issue numbers looked similar at first sight. Reusing the file is definitely OK; sorry for the noise!

**Comment by timvandermeij** [comment]
Opening the `issue12392.pdf` file in the preview above or locally with this code checked out, it doesn't print anything I enter in the fields anymore for me while this works with https://mozilla.github.io/pdf.js/web/viewer.html. Could you check why this happens?

**Comment by calixteman** [comment]
In order to see the "effect" of this patch on file `issue12392.pdf`, it's required to have https://github.com/mozilla/pdf.js/pull/12292 before.

**Comment by timvandermeij** [comment]
Could you rebase this one now that #12292 is merged so we can also look into landing this one?

**Review by timvandermeij — APPROVED** [review]

**Comment by timvandermeij** [comment]
Thank you for fixing this!

/botio makeref
